internal/controller/bucket_controller.go

```diff
@@ -840,17 +861,45 @@ func (r *BucketReconciler) setupCredentials(ctx context.Context, obj *sourcev1.B
 func (r *BucketReconciler) createBucketProvider(ctx context.Context, obj *sourcev1.Bucket, creds *bucketCredentials) (BucketProvider, error) {
 	switch obj.Spec.Provider {
 	case sourcev1.BucketProviderGoogle:
-		if err := gcp.ValidateSecret(creds.secret); err != nil {
-			return nil, err
-		}
 		var opts []gcp.Option
-		if creds.secret != nil {
-			opts = append(opts, gcp.WithSecret(creds.secret))
-		}
 		if creds.proxyURL != nil {
 			opts = append(opts, gcp.WithProxyURL(creds.proxyURL))
 		}
-		return gcp.NewClient(ctx, opts...)
+
+		if creds.secret != nil {
+			if err := gcp.ValidateSecret(creds.secret); err != nil {
+				return nil, err
+			}
+			opts = append(opts, gcp.WithSecret(creds.secret))
+		} else {
+			var authOpts []auth.Option
+
+			if obj.Spec.ServiceAccountName != "" {
+				serviceAccount := client.ObjectKey{
+					Name:      obj.Spec.ServiceAccountName,
+					Namespace: obj.GetNamespace(),
+				}
+				authOpts = append(authOpts, auth.WithServiceAccount(serviceAccount, r.Client))
+			}
+
+			if r.TokenCache != nil {
+				involvedObject := cache.InvolvedObject{
+					Kind:      sourcev1.BucketKind,
+					Name:      obj.GetName(),
+					Namespace: obj.GetNamespace(),
+					Operation: cache.OperationReconcile,
+				}
+				authOpts = append(authOpts, auth.WithCache(*r.TokenCache, involvedObject))
+			}
+
+			if creds.proxyURL != nil {
+				authOpts = append(authOpts, auth.WithProxyURL(*creds.proxyURL))
+			}
```

Comment:
We could move this entire block to right before the `switch obj.Spec.Provider {` above, as all the three providers will use it. This will reduce the amount of changes needed in the next PR, be it `aws` or `azure`